server/graphql/loaders/comments.ts

```diff
@@ -6,18 +6,22 @@ import { ReactionEmoji } from '../../constants/reaction-emoji';
 import models, { Op, sequelize } from '../../models';
 import EmojiReaction from '../../models/EmojiReaction';
 
-import { sortResults } from './helpers';
+type CommentCountByExpenseIdAndType = {
+  ExpenseId: number;
+  type: string | string[];
+};
+
 export default {
-  countByExpenseId: (): DataLoader<number, number> =>
-    new DataLoader(ExpenseIds =>
-      models.Comment.count({
+  countByExpenseAndType: (): DataLoader<CommentCountByExpenseIdAndType, number> =>
+    new DataLoader(async (ExpenseIdAndType: Array<CommentCountByExpenseIdAndType>) => {
+      const counters = await models.Comment.count({
         attributes: ['ExpenseId'],
-        where: { ExpenseId: { [Op.in]: ExpenseIds } },
+        where: { [Op.or]: ExpenseIdAndType },
         group: ['ExpenseId'],
-      }).then(results =>
-        sortResults(ExpenseIds, results, 'ExpenseId', { count: 0 }).map(result => <number>result.count),
-      ),
-    ),
+      });
+      return ExpenseIdAndType.map(({ ExpenseId }) => counters.find(c => c.ExpenseId === ExpenseId)?.count || 0);
+    }),
+
```

Comment:
If not passing primitives (numbers, strings) that `dataloader` can natively cache on, you need to provide a `cacheKeyFn`. Otherwise it won't be able to group or cache the data properly.